packages/store/ts/config/v2/table.ts

```diff
@@ -1,99 +1,184 @@
+    Array.isArray(key) &&
+    key.every(
+      (key) =>
+        hasOwnKey(schema, key) && hasOwnKey(scope.types, schema[key]) && isStaticAbiType(scope.types[schema[key]]),
+    )
+  );
+}
+
+/** @deprecated */
+export function isTableInput(input: unknown): input is TableInput {
+  return (
+    typeof input === "object" &&
+    input !== null &&
+    hasOwnKey(input, "schema") &&
+    hasOwnKey(input, "key") &&
+    Array.isArray(input["key"])
+  );
+}
+
+export type validateKeys<validKeys extends PropertyKey, keys> = {
+  [i in keyof keys]: keys[i] extends validKeys ? keys[i] : validKeys;
+};
+
+export type validateTable<input, scope extends Scope = AbiTypeScope> = {
+  [key in keyof input]: key extends "key"
+    ? validateKeys<getStaticAbiTypeKeys<conform<get<input, "schema">, SchemaInput>, scope>, input[key]>
+    : key extends "schema"
+      ? validateSchema<input[key], scope>
+      : key extends "name" | "namespace"
+        ? narrow<input[key]>
+        : key extends keyof TableInput
+          ? TableInput[key]
+          : ErrorMessage<`Key \`${key & string}\` does not exist in TableInput`>;
+};
+
+export function validateTable<input, scope extends Scope = AbiTypeScope>(
+  input: input,
+  scope: scope = AbiTypeScope as unknown as scope,
+): asserts input is TableInput & input {
+  if (typeof input !== "object" || input == null) {
+    throw new Error(`Expected full table config, got \`${JSON.stringify(input)}\``);
+  }
+
+  if (!hasOwnKey(input, "schema")) {
+    throw new Error("Missing schema input");
   }
-  validateTableFull(input, scope);
+  validateSchema(input.schema, scope);
+
+  if (!hasOwnKey(input, "key") || !isValidPrimaryKey(input["key"], input["schema"], scope)) {
+    throw new Error(
+      `Invalid key. Expected \`(${getValidKeys(input["schema"], scope)
+        .map((item) => `"${String(item)}"`)
+        .join(" | ")})[]\`, received \`${
+        hasOwnKey(input, "key") && Array.isArray(input.key)
+          ? `[${input.key.map((item) => `"${item}"`).join(", ")}]`
+          : "undefined"
+      }\``,
+    );
+  }
+}
+
+export type resolveTableCodegen<input extends TableInput> = {
+  [key in keyof TableCodegen]-?: key extends keyof input["codegen"]
+    ? undefined extends input["codegen"][key]
+      ? key extends "dataStruct"
+        ? boolean
+        : key extends keyof typeof TABLE_CODEGEN_DEFAULTS
+          ? (typeof TABLE_CODEGEN_DEFAULTS)[key]
+          : never
+      : input["codegen"][key]
+    : // dataStruct isn't narrowed, because its value is conditional on the number of value schema fields
+      key extends "dataStruct"
+      ? boolean
+      : key extends keyof typeof TABLE_CODEGEN_DEFAULTS
+        ? (typeof TABLE_CODEGEN_DEFAULTS)[key]
+        : never;
+};
+
+export function resolveTableCodegen<input extends TableInput>(input: input): resolveTableCodegen<input> {
+  const options = input.codegen;
+  return {
+    directory: get(options, "directory") ?? TABLE_CODEGEN_DEFAULTS.directory,
```

Comment:
I wonder if we should name this something like `output` or `outputDirectory` or `outDir`